internal/locate/region_cache.go

```diff
@@ -1533,17 +1623,29 @@ func (c *RegionCache) OnRegionEpochNotMatch(bo *retry.Backoffer, ctx *RPCContext
 		}
 	}
 
+	var buckets *metapb.Buckets
+	c.mu.Lock()
+	cachedRegion, ok := c.mu.regions[ctx.Region]
+	if ok {
+		buckets = cachedRegion.getStore().buckets
+	}
+	c.mu.Unlock()
+
 	needInvalidateOld := true
 	newRegions := make([]*Region, 0, len(currentRegions))
 	// If the region epoch is not ahead of TiKV's, replace region meta in region cache.
 	for _, meta := range currentRegions {
 		if _, ok := c.pdClient.(*CodecPDClient); ok {
 			var err error
+			// Can't modify currentRegions in this function because it can be shared by
+			// multiple goroutines, refer to https://github.com/pingcap/tidb/pull/16962.
 			if meta, err = decodeRegionMetaKeyWithShallowCopy(meta); err != nil {
 				return false, errors.Errorf("newRegion's range key is not encoded: %v, %v", meta, err)
 			}
 		}
-		region, err := newRegion(bo, c, &pd.Region{Meta: meta})
+		// TODO(youjiali1995): new regions inherit old region's buckets now. Maybe we should make EpochNotMatch error
+		// carry buckets information. Can it bring much overhead?
```

Comment:
Is is true that bucket information usually doesn't change greatly just after a region merge or split? Then, an asynchronous `UpdateBucketsIfNeeded` from PD may be ok.

Reply:
Agree. Buckets is generated asynchronously after region leader or range changed. Querying PD immediately may not get the newer buckets, so not handling it and depending on https://github.com/pingcap/tidb/pull/32818/files#diff-23eb1043649369d5956b123b173b799ded8c6b7cbf4519bb32cb3caa01928d75R936 is also a choice.